server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/ExportService.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.james.blob.export.api.BlobExportMechanism;
+import org.apache.james.blob.export.api.FileExtension;
+import org.apache.james.core.Username;
+import org.apache.james.mailbox.backup.MailboxBackup;
+import org.apache.james.task.Task;
+import org.apache.james.user.api.UsersRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class ExportService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExportService.class);
+    private static final String EXPLANATION = "The backup of your mailboxes has been exported to you";
+    private static final String FILE_PREFIX = "mailbox-backup-";
+    public static final boolean LAZY_RESOURCE_CLEANUP = false;
+
+    private final MailboxBackup mailboxBackup;
+    private final BlobStore blobStore;
+    private final BlobExportMechanism blobExport;
+    private final UsersRepository usersRepository;
+
+    public ExportService(MailboxBackup mailboxBackup, BlobStore blobStore, BlobExportMechanism blobExport, UsersRepository usersRepository) {
+        this.mailboxBackup = mailboxBackup;
+        this.blobStore = blobStore;
+        this.blobExport = blobExport;
+        this.usersRepository = usersRepository;
+    }
+
+    public Mono<Task.Result> exportUserMailboxesData(Username username) {
+        return Mono.using(
+            () -> zipMailboxesContent(username),
+            inputStream -> blobStore.save(blobStore.getDefaultBucketName(), inputStream, BlobStore.StoragePolicy.LOW_COST)
+                .flatMap(blobId -> export(username, blobId)
+                    .doFinally(any -> deleteBlob(blobId))
+                    .thenReturn(Task.Result.COMPLETED))
+                .onErrorResume(e -> {
+                    LOGGER.error("Error exporting mailboxes of user: {}", username.asString(), e);
+                    return Mono.just(Task.Result.PARTIAL);
+                }),
+            Throwing.consumer(InputStream::close).sneakyThrow(),
+            LAZY_RESOURCE_CLEANUP);
+    }
+
+    private Mono<Void> export(Username username, BlobId blobId) {
+        return Mono.fromRunnable(Throwing.runnable(() ->
+            blobExport.blobId(blobId)
+                .with(usersRepository.getMailAddressFor(username))
+                .explanation(EXPLANATION)
+                .filePrefix(FILE_PREFIX + username.asString() + "-")
+                .fileExtension(FileExtension.ZIP)
+                .export()));
+    }
+
+    private void deleteBlob(BlobId blobId) {
+        blobStore.delete(blobStore.getDefaultBucketName(), blobId)
+            .onErrorResume(e -> {
+                LOGGER.error("Error deleting Blob with blobId: {}", blobId.asString(), e);
+                return Mono.empty();
+            }).block();
+    }
+
+    private InputStream zipMailboxesContent(Username username) throws Exception {
```

Comment:
It's a good idea but I really want that to be a general purpose tool with the test coverage it deserves.

Reply:
What happen when I close the returned InputStream? Does it close the subscriber properly?

Reply:
I'm not sure. If we close the InputStream while a subscriber is reading from it then an IOException will be thrown, and we end up in the onErrorResume block? Or did you mean the subscriber in the zipMailboxesContent itself?